### PR TITLE
Sign up reactive forms

### PR DESCRIPTION
**Pull request by miguelhenaoa**
# Summary
- Change FormsModule to ReactiveFormsModule
- Add validators match password
- Add validators required for all fields,  pattern for password field and password_confirm

# Trello Card
https://trello.com/c/OpIkB1oP/5-angular-reactive-forms

# Screenshot
Captura de Pantalla 2021-05-25 a la(s) 1 42 27 p  m
Captura de Pantalla 2021-05-25 a la(s) 1 43 42 p  m

### CODE REVIEW

File: src/app/screens/sign-up/sign-up.component.ts [hunk 1]
```diff
@@ -1,18 +1,64 @@
...
+  }
+
+  isControlHasError(controlName: string, validationType: string): boolean {
+    const control = this.form.get(controlName);
+    if (!control) {
+      return false;
+    }
+
+    const result = control.hasError(validationType) && (control.dirty || control.touched);
+    return result;
```
Comment (ebedoyaalzate) [inline src/app/screens/sign-up/sign-up.component.ts]: You could refactor this method with conditional ternary operator.

Comment (ebedoyaalzate) [inline src/app/screens/sign-up/sign-up.component.ts:49]: You could refactor this method with conditional ternary operator.

File: src/app/screens/sign-up/sign-up.component.ts [hunk 2]
```diff
@@ -1,18 +1,64 @@
...
+  }
+
+  signUp(): void {
+    if (this.form.valid) {
+      console.log(this.form.value);
+    }
+  }
+
+  validateForm(control: string, form: FormGroup, validationType: string): boolean {
```
Comment (ebedoyaalzate) [inline src/app/screens/sign-up/sign-up.component.ts]: isControlHasError should work for this case.

Comment (ebedoyaalzate) [inline src/app/screens/sign-up/sign-up.component.ts:58]: isControlHasError should work for this case.

**Review by ebedoyaalzate — CHANGES_REQUESTED** [review]

File: src/app/screens/sign-up/sign-up.component.html [hunk 3]
```diff
@@ -1,9 +1,20 @@
...
+      <div class="alert">
+        <small *ngIf="isControlHasError(field.formControlName, 'required')">{{field.label}} es requerido.</small>
+        <small *ngIf="isControlHasError(field.formControlName, 'minLength')">El campo {{field.label}} debe tener mínimo 3 caracteres.</small>
+        <small *ngIf="isControlHasError(field.formControlName, 'email')">Formato del correo inválido.</small>
+        <small *ngIf="isControlHasError(field.formControlName, 'pattern') && field.formControlName === 'password'">
+          La contraseña debe tener mínimo una mayúscula, una minúscula y un número.
+        </small>
+        <small *ngIf="isControlHasError(field.formControlName, 'MatchPassword')">
+          Las contraseñas no coinciden. 
+        </small>
```
Comment (ebedoyaalzate) [inline src/app/screens/sign-up/sign-up.component.html]: You could refactor this part just let one small tag and get the message from ts

File: src/app/screens/sign-up/sign-up.component.html [hunk 4]
```diff
@@ -1,9 +1,20 @@
 <app-form [buttonLabel]="'Login'" [buttonRoute]="'/login'">
-  <form class="content-form" (ngSubmit)="signUp()">
+  <form class="content-form" [formGroup]="form" (ngSubmit)="signUp()">
     <div class="content-input" *ngFor="let field of fields">
       <label [for]="field.name">{{field.label}}</label>
-      <input class="input-base" [type]="field.type" [(ngModel)]="signUpForm[field.name]" [ngModelOptions]="{ standalone: true}">
+      <input class="input-base" [type]="field.type" [formControlName]="field.formControlName">
+      <div class="alert">
```
Comment (ebedoyaalzate) [inline src/app/screens/sign-up/sign-up.component.html]: This div isn't necessary

File: src/app/screens/sign-up/sign-up.component.scss [hunk 5]
```diff
@@ -16,3 +20,8 @@
   flex-flow: column;
   margin: 15px 0;
 }
+
+.disabled {
+  background-color: $silver-chalice;
+  cursor: default;
+}
```
Comment (ebedoyaalzate) [inline src/app/screens/sign-up/sign-up.component.scss]: Add empty line.

**Review by ebedoyaalzate — CHANGES_REQUESTED** [review]

File: src/app/screens/sign-up/sign-up.component.ts [hunk 6]
```diff
@@ -1,18 +1,78 @@
...
+      } else if (control?.hasError('minLength')) {
+        field.errorMessage = `El campo ${field.label} debe tener mínimo 3 caracteres.`;
+        return true;
+      } else if (control?.hasError('email')) {
+        field.errorMessage = `Formato del correo inválido.`;
+        return true;
+      } else if (control?.hasError('pattern')) {
+        field.errorMessage = `La contraseña debe tener mínimo una mayúscula, una minúscula y un número.`;
+        return true;
+      }
```
Comment (ebedoyaalzate) [inline src/app/screens/sign-up/sign-up.component.ts]: You could refactor this if/else structure.

**Review by ebedoyaalzate — CHANGES_REQUESTED** [review]

File: src/app/screens/sign-up/sign-up.component.ts [hunk 7]
```diff
@@ -1,18 +1,71 @@
...
+      { validators: validatePasswordForm() }
+    );
+  }
+
+  isControlHasError(field: any): boolean {
+    const controlName = field.formControlName;
+    const control = this.form.get(controlName);
+
+    if (controlName === 'password_confirmation') {
+      if (this.form.errors?.['MatchPassword']) {
```
Comment (ebedoyaalzate) [inline src/app/screens/sign-up/sign-up.component.ts]: Use && operator

**Review by ebedoyaalzate — CHANGES_REQUESTED** [review]

**Review by ebedoyaalzate — APPROVED** [review]

File: src/app/screens/sign-up/sign-up.component.ts [hunk 8]
```diff
@@ -1,18 +1,69 @@
...
+          field.errorMessage = validator.errorMessage;
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  signUp(): void {
+    if (this.form.valid) {
```
Comment (CrisVega08) [inline src/app/screens/sign-up/sign-up.component.ts:64]: I think this conditional is unnecessary, you are validating the form before click event

**Review by CrisVega08 — APPROVED** [review]

**Pull request merged**